src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,322 @@
+                                                         std::string_view module,
+                                                         std::string_view fallback_function,
+                                                         std::string_view fallback_module) {
+  ErrorMessageOr<uint64_t> primary_address_or_error = FindFunctionAddress(pid, function, module);
+  if (primary_address_or_error.has_value()) {
+    return primary_address_or_error.value();
+  }
+  ErrorMessageOr<uint64_t> fallback_address_or_error =
+      FindFunctionAddress(pid, fallback_function, fallback_module);
+  if (fallback_address_or_error.has_value()) {
+    return fallback_address_or_error.value();
+  }
+
+  return ErrorMessage(absl::StrFormat(
+      "Failed to load symbol \"%s\" from module \"%s\" with error: \"%s\"\nAnd also "
+      "failed to load fallback symbol \"%s\" from module \"%s\" with error: \"%s\"",
+      function, module, primary_address_or_error.error().message(), fallback_function,
+      fallback_module, fallback_address_or_error.error().message()));
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void*> DlopenInTracee(pid_t pid, std::filesystem::path path,
+                                                   uint32_t flag) {
+  // Figure out address of dlopen in libc.
+  OUTCOME_TRY(address_dlopen, FindFunctionAddressWithFallback(pid, "dlopen", "libdl-",
+                                                              "__libc_dlopen_mode", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee. This is used for the code and the path name.
+  const uint64_t path_length = path.string().length() + 1;  // Include terminating zero.
+  const uint64_t memory_size = kCodeScratchPadSize + path_length;
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, memory_size));
+
+  // Write the name of the .so into memory at address_code with offset of kCodeScratchPadSize.
+  std::vector<uint8_t> path_as_vector(path_length);
+  memcpy(path_as_vector.data(), path.c_str(), path_length);
+  const uint64_t address_so_path = address_code + kCodeScratchPadSize;
+  auto result_write_path = WriteTraceesMemory(pid, address_so_path, path_as_vector);
+  if (result_write_path.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_path.error();
+  }
+
+  // We want to do the following in the tracee:
+  // return_value = dlopen(path, flag);
+  // The calling convention is to put the parameters in registers rdi and rsi.
+  // So the address of the file path goes to rdi. The flag argument goes into rsi. Then we load the
+  // address of dlopen into rax and do the call. Assembly in Intel syntax (destination first),
+  // machine code on the right:
+
+  // movabs rax, address_so_path      48 b8 address_so_path
+  // mov rbx, 2                       48 c7 c3 flag
```

Comment:
So by now I get that you cannot do `movabs rdi address_so_path`, but can't you do `mov rsi 2` instead of going through `rbx`?

Reply:
Okay, I digged deeper into this and it turns out the stuff I claimed above was rubbish: One can move immediate 64 bit values into all registers. The same is true for 32 bit values. The "2" in the comment needs to be changed to "flag".
I changed that here and also in the other places where I write machine code directly.

Reply:
Nice, much cleaner.